content/blog/typescript-type-generics/index.md

```diff
@@ -0,0 +1,340 @@
+    return returnProp;
+}
+
+const returnedObject = returnSelf({objProperty: 12}); // This now works! 🎉
+
+returnedObject.test(); // This will not return an error but should 🙁
+returnedObject.objProperty; // This will also (correctly) not throw an error, but TS won't know it's a number ☹️
+```
+
+# The Real Solution {#generics-intro}
+
+So what's the answer? How can we get preserved type data on both the input and the output??
+
+The solution is... Well, you've read the title I'm sure.
+
+_Type generics allow us to store loose type data in a **type variable**_. A type variable is _a unique kind of variable that's not exposed to JavaScript but is instead handled by TypeScript to provide expected typing data_. For example, the above example could be rewritten as:
+
+```typescript
+function returnSelf<T>(returnProp: T): T {
+    return returnProp;
+}
+```
+
+In this example, we're defining a type variable `T`, then telling TS that both the property and the return type should be the same type.
+
+This means that you can use the function like this:
+
+```typescript
+const numberVar = returnSelf(2); // T in this instance is `2`, so it's similar to writing `const numberVal: 2 = 2;`
+
+// Likewise, this object is now returned as if it was just placed on the const
+const returnedObject = returnSelf({objProperty: 12});
+
+// This will fail, as expected
+returnedObject.test();
+// This will exist, and TS will know it as a number
+returnedObject.objProperty;
+```
+
+> Author's note:
+>
+> The type variable does not need to be called `T`. In fact, while it seems to be commonplace for the community to use single-letter type variable names (often due to the length and complexity of the typings), there are many reasons why more explicit type names should be used.
+>
+> Remember, type variables are like other variables in that you need to maintain them and understand what they're doing in your code.
+
+# Okay, but Why? {#logger-example}
+
+Why might we want to do this? [Returning an item as itself in a identity function](#generic-usecase-setup) is cool, but it's not very useful in its current state. That said, there ARE many many uses for generics in real-world codebases.
+
+For example, let's say that we had the following JavaScript code that we wanted to use as a logger:
+
+```javascript
+const util = require('util'),
+   	  fs   = require('fs');
+
+// Have the `writeFile` return a promise instead of having to use a callback
+const writeFileAsync = util.promisify(fs.writeFile);
+
+/**
+ * Async functions allows us to use `await` on promises in the function body, try/catch them, and
+ * will return their own promise wrapped around the `return` value
+ */
+async function logTheValue(item) {
+	const jsonString = JSON.stringify(item, null, 2);
+
+	let err = undefined;
+
+	try {
+		// Attempt to write a new log file. If this fails, save the error to the `err` variable
+		await writeFileAsync(`/logs/${Date.now()}`, jsonString);
+	} catch (e) {
+		err = e;
```

Comment:
The manual "hoisting" of `e` to `err` may not be the easiest for some, especially with two shortenings rather than verbose names. Not sure what's best to do here, but in general I like being explicit since `e` is often `event` and `err` saves only two letters.

Reply:
`error` can be preserved in some instances so I wanted to remove it. I'll add a comment to make it more clear but I don't think I want to change the code too much here